tls/src/lib.rs

```diff
@@ -0,0 +1,497 @@
+#[derive(Debug)]
+pub(crate) struct LocalEd25519SigningKey {
+    privkey_der: Secret<PrivkeyDer>,
+}
+
+impl SigningKey for LocalEd25519SigningKey {
+    fn choose_scheme(
+        &self,
+        offered: &[SignatureScheme],
+    ) -> Option<Box<dyn rustls::sign::Signer>> {
+        if !offered.iter().any(|&s| s == SignatureScheme::ED25519) {
+            return None;
+        }
+
+        let privkey_info = PrivateKeyInfo::try_from(
+            self.privkey_der.expose_secret().0.as_slice(),
+        )
+        .ok()?;
+
+        let signing_key =
+            ed25519_dalek::SigningKey::try_from(privkey_info).ok()?;
+
+        Some(Box::new(LocalEd25519Signer { key: signing_key })
+            as Box<dyn Signer>)
+    }
+
+    fn algorithm(&self) -> rustls::SignatureAlgorithm {
+        rustls::SignatureAlgorithm::ED25519
+    }
+}
+
+/// A verifier for certs generated on the RoT
+#[derive(Debug)]
+struct RotCertVerifier {
+    verifier: PkiPathSignatureVerifier,
+    log: slog::Logger,
+}
+
+impl RotCertVerifier {
+    pub fn new(root: Certificate, log: slog::Logger) -> anyhow::Result<Self> {
+        let verifier = PkiPathSignatureVerifier::new(Some(root))?;
+        Ok(RotCertVerifier { verifier, log })
+    }
+
+    /// Create a `PkiPath` suitable for `dice-verifier`
+    fn pki_path(
+        end_entity: &rustls::pki_types::CertificateDer<'_>,
+        intermediates: &[rustls::pki_types::CertificateDer<'_>],
+    ) -> Result<Vec<Certificate>, rustls::Error> {
+        let mut pki_path = Vec::new();
+        for der in iter::once(end_entity).chain(intermediates) {
+            pki_path.push(Certificate::from_der(der).map_err(|_| {
+                rustls::Error::InvalidCertificate(
+                    rustls::CertificateError::BadEncoding,
+                )
+            })?)
+        }
+        Ok(pki_path)
+    }
+
+    /// Verify the certificate chain via `dice-verifier`
+    pub fn verify_cert(
+        &self,
+        end_entity: &rustls::pki_types::CertificateDer<'_>,
+        intermediates: &[rustls::pki_types::CertificateDer<'_>],
+    ) -> Result<(), rustls::Error> {
+        let pki_path = Self::pki_path(end_entity, intermediates)?;
+        self.verifier.verify(&pki_path).map_err(|e| {
+            error!(self.log, "Failed to verify cert: {e}");
+            rustls::Error::InvalidCertificate(
+                rustls::CertificateError::BadEncoding,
+            )
+        })?;
+
+        info!(self.log, "Certificate chain verified successfully");
+
+        Ok(())
+    }
+
+    fn verify_signature(
+        &self,
+        message: &[u8],
+        cert: &rustls::pki_types::CertificateDer<'_>,
+        sig: &[u8],
+    ) -> Result<rustls::client::danger::HandshakeSignatureValid, rustls::Error>
+    {
+        // Get the public key
+        let cert = Certificate::from_der(cert).map_err(|_| {
+            rustls::Error::InvalidCertificate(
+                rustls::CertificateError::BadEncoding,
+            )
```

Comment:
I don't think it really matters. We only use the end entity cert to do verification, and we'd almost certainly have to print the DER to identify it. If we get a BadEncoding, we'll know which device we are talking to and that will be enough to figure out the problem. In that case it will likely be a *major* problem or a bit flip.